Don't suggest things when the entered identifier is in scope

forki opened the pull request:
this is a partial fix for #2214 

When we detect that the identifier itself is in scope, then we don't suggest stuff since there are good chances that something else went wrong.
In this case we don't suggest anything since quickfix tooling interfers with autocomplete.

forki commented:
@vasily-kirichenko interestingly this is something that only happens in editor. The compiler in our test framework only emits the "Missing qualification after '.'" error.

KevinRansom commented on src/fsharp/ErrorResolutionHints.fs:
```diff
@@ -23,9 +23,11 @@ let IsInEditDistanceProximity userEntered suggestion =
-    suggestionF()
+    let allSuggestions = suggestionF()
+
+    if allSuggestions.Contains userEntered then [] else // some other parsing error occurred
```
Shouldn't the IDE take care of this?  It seems like it's more of an rendering thing than something that makes sense here ?  It seems weird to me that the compiler looks for a suggestion with the text user entered in order to determine whether to filter it.

Perhaps I am missing something though

forki commented:
The idea is the following: if our finds the exact identifier in scope that
the name resolver rejects then something is very fucky and we just don't
suggest anything. It clearly is not a typo but an issue with something
else.

The additional test that I fixed here is a similar issue.

Am 10.01.2017 20:00 schrieb "Kevin Ransom (msft)" <notifications@github.com
>:

> *@KevinRansom* approved this pull request.
>
> The code looks fine apart from user-entered, I'm not sure I feel great
> about that functionality.
> ------------------------------
>
> In src/fsharp/ErrorResolutionHints.fs
> <https://github.com/Microsoft/visualfsharp/pull/2218#pullrequestreview-15986961>
> :
>
> > @@ -23,9 +23,11 @@ let IsInEditDistanceProximity userEntered suggestion =
>  /// Filters predictions based on edit distance to the given unknown identifier.
>  let FilterPredictions (userEntered:string) (suggestionF:ErrorLogger.Suggestions) =
>      let uppercaseText = userEntered.ToUpperInvariant()
> -    suggestionF()
> +    let allSuggestions = suggestionF()
> +
> +    if allSuggestions.Contains userEntered then [] else // some other parsing error occurred
>
> Shouldn't the IDE take care of this? It seems like it's more of an
> rendering thing than something that makes sense here ? It seems weird to me
> that the compiler looks for a suggestion with the text user entered in
> order to determine whether to filter it.
>
> Perhaps I am missing something though
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/visualfsharp/pull/2218#pullrequestreview-15986961>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADgNJzsJgyQys4etMWx0U1lney5ZWrAks5rQ9VVgaJpZM4LfRZ1>
> .
>

KevinRansom commented:
What I find a bit yuky was not really added by this PR, I just wondered why the compiler knows about “user entered” in error recovery.  That seems like an IDE specific element … and should be dealt with by the IDE.


Kevin


From: Steffen Forkmann [mailto:notifications@github.com]
Sent: Tuesday, January 10, 2017 11:18 AM
To: Microsoft/visualfsharp <visualfsharp@noreply.github.com>
Cc: Kevin Ransom <Kevin.Ransom@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [Microsoft/visualfsharp] Don't suggest things when the entered identifier is in scope (#2218)

The idea is the following: if our finds the exact identifier in scope that
the name resolver rejects then something is very fucky and we just don't
suggest anything. It clearly is not a typo but an issue with something
else.

The additional test that I fixed here is a similar issue.

Am 10.01.2017 20:00 schrieb "Kevin Ransom (msft)" <notifications@github.com
<mailto:notifications@github.com%0b>>:

> *@KevinRansom* approved this pull request.
>
> The code looks fine apart from user-entered, I'm not sure I feel great
> about that functionality.
> ------------------------------
>
> In src/fsharp/ErrorResolutionHints.fs
> <https://github.com/Microsoft/visualfsharp/pull/2218#pullrequestreview-15986961>
> :
>
> > @@ -23,9 +23,11 @@ let IsInEditDistanceProximity userEntered suggestion =
> /// Filters predictions based on edit distance to the given unknown identifier.
> let FilterPredictions (userEntered:string) (suggestionF:ErrorLogger.Suggestions) =
> let uppercaseText = userEntered.ToUpperInvariant()
> - suggestionF()
> + let allSuggestions = suggestionF()
> +
> + if allSuggestions.Contains userEntered then [] else // some other parsing error occurred
>
> Shouldn't the IDE take care of this? It seems like it's more of an
> rendering thing than something that makes sense here ? It seems weird to me
> that the compiler looks for a suggestion with the text user entered in
> order to determine whether to filter it.
>
> Perhaps I am missing something though
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/visualfsharp/pull/2218#pullrequestreview-15986961>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADgNJzsJgyQys4etMWx0U1lney5ZWrAks5rQ9VVgaJpZM4LfRZ1>
> .
>


—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FMicrosoft%2Fvisualfsharp%2Fpull%2F2218%23issuecomment-271670220&data=02%7C01%7CKevin.Ransom%40microsoft.com%7C78c519b177d84a64f4a808d4398d71cb%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636196727042482800&sdata=%2FukddCUYJEqpX8f7Wn8uALXw%2Bl4EaRwlnEUss55BYdY%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAE76FqYBlxyCm5GnWB2BkznAfHTAF4phks5rQ9l8gaJpZM4LfRZ1&data=02%7C01%7CKevin.Ransom%40microsoft.com%7C78c519b177d84a64f4a808d4398d71cb%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636196727042482800&sdata=DvnUXWZWMfE8nmRVU0C6esVtz55C90Y1ca5LFzrwXZ4%3D&reserved=0>.

forki commented:
Yeah I took that terminology from elm, but I think idText is better in context of F#. I changed that back.

Also I rebased on master and merge conflict is resolved

KevinRansom commented on src/fsharp/CompileOps.fs:
```diff
@@ -610,7 +610,8 @@ let getErrorString key = SR.GetString key
 
 let OutputPhasedErrorR errorStyle (os:System.Text.StringBuilder) (err:PhasedDiagnostic) =
-    let rec OutputExceptionR (os:System.Text.StringBuilder) = function        
+    let rec OutputExceptionR (os:System.Text.StringBuilder) error = 
```
System.Text is opened at the top of the file:  https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/CompileOps.fs#L7

KevinRansom commented on src/fsharp/ErrorResolutionHints.fs:
```diff
@@ -9,26 +9,28 @@ let highConfidenceThreshold = 0.85
-    let editDistance = Internal.Utilities.EditDistance.CalcEditDistance(userEntered,suggestion)
+/// The threshhold is set to about a quarter of the number of characters.
+let IsInEditDistanceProximity idText suggestion =
+    let editDistance = Internal.Utilities.EditDistance.CalcEditDistance(idText,suggestion)
```
Ditto: Internal.Utilities ---https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/CompileOps.fs#L12

KevinRansom commented on src/fsharp/ErrorResolutionHints.fs:
```diff
@@ -9,26 +9,28 @@ let highConfidenceThreshold = 0.85
+    let uppercaseText = idText.ToUpperInvariant()
+    let allSuggestions = suggestionF()
+
+    if allSuggestions.Contains idText then [] else // some other parsing error occurred
```
Here you use contains idText a little later you use endswith '.idText'.
Which is the correct test? 'idText' or '"." + idText'

Does suggestion return UpperInvariant()?

if not then if allSuggestions.Contains idText test is probably wrong.
otherwise perhaps:
let suggestedText = suggestion.ToUpperInvariant() 

is redundant.

forki replied:
one is a set and the other one a string. so I think it's correct

forki commented:
@kevinransom better?

KevinRansom commented:
@forki, 
I like idText better, except It's not really clear to me what what it refers to ... perhaps a comment similar to your description above ... obviously without the curse word, because we would just have to change that to get through our internal scanning processes.

Kevin

forki commented:
That's not a curse word. It's a meme. See http://m.quickmeme.com/img/aa/aadcd968fbae17016d14456cdad396dabc7fa92909fbd87033640b226a20a760.jpg

KevinRansom commented:
Policheck is very literal.  It hasn't been introduced to meme's yet.

forki commented:
fixed the namespaces. this is now ready

forki commented:
also fixes #2217 

![image](https://cloud.githubusercontent.com/assets/57396/21856716/43df0c92-d823-11e6-8223-7288aaaaca57.png)
